File: server/model/Order.js
'use strict';

let bookshelf = require('./bookshelf');
let DateUtil  = require('../utils/DateUtil');

const STATUS_PEDDING = 0; //未支付
const STATUS_PAYED   = 1; //已支付

let Order = bookshelf.Model.extend({
    tableName: 'order',
    // user: function() {
    //     return this.belongsTo(User);
    // }
}, {
    /*
     * 总的订单数
     */
    getTotalOrderCount: () => {
        return Order.count();
    },

    /*
     * 指定日期的订单数
     */
    getOrderCountByDate: (date) => {
        let start    = new Date(date.year, date.month - 1, date.date).getTime();
        let tomorrow = start + 24 * 60 * 60 * 1000;
        start        = DateUtil.longToYmdStr(start);
        tomorrow     = DateUtil.longToYmdStr(tomorrow);
        return Order.where('created_at', '>=', start)
                    .where('created_at', '<',  tomorrow).count();
    },

    /*
     * 指定日期的销售额
     */
    getTotalSaleByDate: (date) => {
        return new Promise(async function(resolve, reject) {
            let start    = new Date(date.year, date.month - 1, date.date).getTime();
            let tomorrow = start + 24 * 60 * 60 * 1000;
            start        = DateUtil.longToYmdStr(start);
            tomorrow     = DateUtil.longToYmdStr(tomorrow);
            try {
                let sum = await Order.query('sum', 'payment as totalPay')
                            .where('pay_at', '>=', start)
                            .where('pay_at', '<',  tomorrow)
                            .where({
                                status: STATUS_PAYED
                            })
                            .fetch();
                resolve(sum.get('totalPay'));
            } catch (err) {
                reject(err);
            }
        });
    },

    /*
     * 总的销售额
     */
    getTotalSale: () => {
        return new Promise(async function(resolve, reject) {
            try {
                let sum = await Order.query('sum', 'payment as totalPay')
                                    .where({
                                        status: STATUS_PAYED
                                    })
                                    .fetch();
                resolve(sum.get('totalPay'));
            } catch (err) {
                reject(err);
            }
        });
    },

    /*
     * 近30天，每天的销售额
     */
    getSaleFor30d: async () => {
        let today    = new Date();
        today        = new Date(today.getFullYear(), today.getMonth(), today.getDate());
        let before29 = today.getTime() - 29 * 24 * 60 * 60 * 1000;
        before29     = DateUtil.longToYmdStr(before29);
        let status   = STATUS_PAYED;
        var sql = `
            SELECT sum(payment) as amount, DATE_FORMAT(pay_at,'%Y-%m-%d') as payAt
            FROM \`order\`
            WHERE pay_at > ? and status = ?
            GROUP BY DATE_FORMAT(pay_at,'%Y-%m-%d');
        `;
        try {
            let result = await bookshelf.knex.raw(sql, [before29, status]);
            return result[0];
        } catch (err) {
            throw err;
        }
    },

    /*
     * 近30天，每天的订单数
     */
    getOrderFor30d: async () => {
        let today    = new Date();
        today        = new Date(today.getFullYear(), today.getMonth(), today.getDate());
        let before29 = today.getTime() - 29 * 24 * 60 * 60 * 1000;
        before29     = DateUtil.longToYmdStr(before29);
        var sql = `
            SELECT count(id) as count, DATE_FORMAT(created_at,'%Y-%m-%d') as createdAt
            FROM \`order\`
            WHERE created_at > ?
            GROUP BY DATE_FORMAT(created_at,'%Y-%m-%d');
        `;
        try {
            let result = await bookshelf.knex.raw(sql, [before29]);
            return result[0];
        } catch (err) {
            throw err;
        }
    }
});

module.exports = Order;

File: server/model/order.go
package model

import (
	"strings"
	"time"
	"../config"
	"github.com/jinzhu/gorm"
)

// Order 订单
type Order struct {
    ID             uint       `gorm:"primary_key" json:"id"`
	CreatedAt      time.Time  `json:"createdAt"`
	UpdatedAt      time.Time  `json:"updatedAt"`
	DeletedAt      *time.Time `sql:"index" json:"deletedAt"`
    UserID         uint       `json:"userId"`
    TotalPrice     float64    `json:"totalPrice"`
    Payment        float64    `json:"payment"`
    Freight        float64    `json:"freight"`
    Remark         string     `json:"remark"`      
    Discount       int        `json:"discount"`
    DeliverStart   time.Time  `json:"deliverStart"`
    DeliverEnd     time.Time  `json:"deliverEnd"`
    Status         int        `json:"status"`
	PayAt          time.Time  `json:"payAt"`
}

// OrderStatusPending 未支付
const OrderStatusPending  = 0

// OrderStatusPaid 已支付
const OrderStatusPaid = 1


// OrderPerDay 每天的订单数
type OrderPerDay []struct {
	Count        int    `json:"count"`
	CreatedAt    string `gorm:"column:createdAt" json:"createdAt"` 
}

// Latest30Day 近30天，每天的订单数
func (orders OrderPerDay) Latest30Day() (OrderPerDay) {
    now          := time.Now()
	year         := now.Year()
	month        := now.Month()
	date         := now.Day()
	today        := time.Date(year, month, date, 0, 0, 0, 0, time.Local)

	before29     := today.Unix() - 29 * 24 * 60 * 60; //29天前（秒）
	before29Date := time.Unix(before29, 0)

	sqlData      := before29Date.Format("2006-01-02")
	sqlArr       := []string{
		"SELECT count(id) as count, DATE_FORMAT(created_at,'%Y-%m-%d') as createdAt",
		"FROM `orders`",
		"WHERE created_at > ?",
		"GROUP BY DATE_FORMAT(created_at,'%Y-%m-%d');",
	}
	sql     := strings.Join(sqlArr, " ")
	db, err := gorm.Open(config.DBConfig.Dialect, config.DBConfig.URL)
	if err != nil {
		return nil
	}

	var result OrderPerDay
	err = db.Raw(sql, sqlData).Scan(&result).Error
	if err != nil {
		return nil
	}
	return result
}


// AmountPerDay 每天的销售额
type AmountPerDay []struct {
	Amount   float64 `json:"amount"`
	PayAt    string  `gorm:"column:payAt" json:"payAt"` 
}

// AmountLatest30Day 近30天，每天的销售额
func (amount AmountPerDay) AmountLatest30Day() (AmountPerDay) {
	now          := time.Now()
	year         := now.Year()
	month        := now.Month()
	date         := now.Day()
	today        := time.Date(year, month, date, 0, 0, 0, 0, time.Local)

	before29     := today.Unix() - 29 * 24 * 60 * 60; //29天前（秒）
	before29Date := time.Unix(before29, 0)

	sqlData      := before29Date.Format("2006-01-02")
	sqlArr       := []string{
		"SELECT sum(payment) as amount, DATE_FORMAT(pay_at,'%Y-%m-%d') as payAt",
		"FROM `orders`",
		"WHERE pay_at > ? and status = ?",
		"GROUP BY DATE_FORMAT(pay_at,'%Y-%m-%d');",
	};

	sql     := strings.Join(sqlArr, " ")
	db, err := gorm.Open(config.DBConfig.Dialect, config.DBConfig.URL)
	if err != nil {
		return nil
	}

	var result AmountPerDay
	err = db.Raw(sql, sqlData, OrderStatusPaid).Scan(&result).Error
	if err != nil {
		return nil
	}
	return result
}